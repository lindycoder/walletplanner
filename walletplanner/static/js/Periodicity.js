import { moment } from './utils';

class Periodicity {
    currentPeriod(date) {
        return this.getPeriod(date);
    }

    * pastPeriods(date) {
        let current = this.currentPeriod(date);
        while (true) {
            current = this.getPeriod(moment(current.start).add(-1, 'days'));
            yield current;
        }
    }

    * futurePeriods(date) {
        let current = this.currentPeriod(date);
        while (true) {
            current = this.getPeriod(moment(current.end).add(1, 'days'));
            yield current;
        }
    }
}

class StaticMonthly extends Periodicity {
    constructor(days) {
        super();
        this.days = days;
    }

    getPeriod(date) {
        let nextPeriodDayIndex = this._nextPeriodDayIndex(date);

        return moment.range(
            this._startDate(nextPeriodDayIndex, date),
            this._endDate(nextPeriodDayIndex, date),
        )
    }

    _nextPeriodDayIndex(date) {
        let nextPeriodDayIndex = 0;
        while (this.days[nextPeriodDayIndex] <= date.date()) {
            nextPeriodDayIndex += 1;
        }
        return nextPeriodDayIndex
    }

    _startDate(nextPeriodDayIndex, date) {
        date = moment(date);
        if (nextPeriodDayIndex === 0) {
            date.add(-1, 'month');
            date.date(this.days[this.days.length - 1])
        } else {
            date.date(this.days[nextPeriodDayIndex - 1])
        }
        return date;
    }

    _endDate(nextPeriodDayIndex, date) {
        date = moment(date);
        if (nextPeriodDayIndex >= this.days.length) {
            date.add(1, 'month')
            date.date(this.days[nextPeriodDayIndex % this.days.length])
        } else {
            date.date(this.days[nextPeriodDayIndex])
        }
        return date;
    }
}

export {
    Periodicity,
    StaticMonthly
}