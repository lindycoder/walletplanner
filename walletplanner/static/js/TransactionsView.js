import React from 'react';
import PropTypes from 'prop-types';
import TransactionsPeriodView from './TransactionsPeriodView';
import Period from "./Period";


export default class TransactionsView extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        let currentPeriod = this.props.periodicity.currentPeriod(this.props.currentDate);

        let displayedPeriods = [];

        let pastGen = this.props.periodicity.pastPeriods(this.props.currentDate);
        for (let i = this.props.periodStart; i < 0; i++) {
            displayedPeriods.unshift(pastGen.next().value);
        }

        displayedPeriods.push(currentPeriod);

        let futureGen = this.props.periodicity.futurePeriods(this.props.currentDate);
        for (let i = 0; i < 5; i++) {
            displayedPeriods.push(futureGen.next().value);
        }

        return (
            <div className="container">
                <div className="row">
                    {displayedPeriods.map((period, periodIndex) => (
                        <TransactionsPeriodView
                            key={periodIndex}
                            className="transaction-view-period col-sm"
                            period={new Period({
                                openingBalance: 0,
                                range: period,
                                transactions: this.props.transactions.filter(
                                    transaction => period.contains(transaction.date, {excludeEnd: true})
                                )
                            })}
                        />
                    ))}
                </div>
            </div>
        );
    }
}


TransactionsView.propTypes = {
    transactions: PropTypes.any.isRequired,
    periodicity: PropTypes.any.isRequired,
    currentDate: PropTypes.any.isRequired,
    periodStart: PropTypes.any.isRequired
};
