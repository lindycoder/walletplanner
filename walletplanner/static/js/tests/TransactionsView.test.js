import React from 'react';
import {shallow} from 'enzyme';
import TransactionsView from '../TransactionsView';
import {StaticMonthly} from '../Periodicity';
import {moment} from '../utils';

describe('TransactionsView', () => {
    let props;
    let wrapper;

    describe('With transactions', () => {

        beforeEach(() => {
            props = {
                transactions: [
                    {"amount": 1000, "date": moment("2018-12-10"), "description": "Before period start", "category": null},
                    {"amount": 1000, "date": moment("2018-12-31"), "description": "3 periods ago", "category": null},
                    {"amount": 1000, "date": moment("2019-01-01"), "description": "2 periods ago", "category": null},
                    {"amount": 1000, "date": moment("2019-01-15"), "description": "1 period ago", "category": null},
                    {"amount": 1000, "date": moment("2019-02-06"), "description": "Current period 1", "category": null},
                    {"amount": 1000, "date": moment("2019-02-08"), "description": "Current period 2", "category": null},
                    {"amount": 1000, "date": moment("2019-02-26"), "description": "Next period 2", "category": null},
                    {"amount": 1000, "date": moment("2019-04-30"), "description": "In 5 periods", "category": null},
                    {"amount": 1000, "date": moment("2019-05-03"), "description": "In 6 periods", "category": null},
                ],
                periodicity: new StaticMonthly([1, 15]),
                currentDate: moment("2019-02-01"),
                periodStart: -3,
                openingBalance: 1000
            };
            wrapper = shallow(<TransactionsView {...props} />);
        });

        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly amounts in the correct column', () => {
            let periods = wrapper.find("TransactionsPeriodView");
            expect(periods).toHaveLength(9, "Periods should be 3 before, current plus 5 next");

            expect(periods.at(0).props().period.transactions).toEqual([props.transactions[1]]);
            expect(periods.at(1).props().period.transactions).toEqual([props.transactions[2]]);
            expect(periods.at(2).props().period.transactions).toEqual([props.transactions[3]]);
            expect(periods.at(3).props().period.transactions).toEqual([props.transactions[4], props.transactions[5]]);
            expect(periods.at(4).props().period.transactions).toEqual([props.transactions[6]]);
            expect(periods.at(5).props().period.transactions).toEqual([]);
            expect(periods.at(6).props().period.transactions).toEqual([]);
            expect(periods.at(7).props().period.transactions).toEqual([]);
            expect(periods.at(8).props().period.transactions).toEqual([props.transactions[7]]);

            expect(periods.at(8).props().period.getClosingBalance()).toBe(9000);
        });
    });

    describe('Without transactions', () => {

        beforeEach(() => {
            props = {
                transactions: [],
                periodicity: new StaticMonthly([1, 15]),
                currentDate: moment("2019-02-01"),
                periodStart: -3,
                openingBalance: 1000
            };
            wrapper = shallow(<TransactionsView {...props} />);
        });

        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly amounts in the correct column', () => {
            let periods = wrapper.find("TransactionsPeriodView");
            expect(periods).toHaveLength(9, "Periods should be 3 before, current plus 5 next");
        });
    });

    describe('Without only 1 transaction in current period', () => {

        beforeEach(() => {
            props = {
                transactions: [
                    {"amount": 1000, "date": moment("2019-02-06"), "description": "In 5 periods", "category": null}
                ],
                periodicity: new StaticMonthly([1, 15]),
                currentDate: moment("2019-02-01"),
                periodStart: -3,
                openingBalance: 1000
            };
            wrapper = shallow(<TransactionsView {...props} />);
        });

        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
