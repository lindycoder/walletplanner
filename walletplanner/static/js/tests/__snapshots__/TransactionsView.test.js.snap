// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsView should render correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="0"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-01-01T00:00:00.000Z",
            "start": "2018-12-15T00:00:00.000Z",
          },
          "transactions": Array [
            Object {
              "amount": 1000,
              "category": null,
              "date": "2018-12-31T00:00:00.000Z",
              "description": "3 periods ago",
            },
          ],
        }
      }
    />
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="1"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-01-15T00:00:00.000Z",
            "start": "2019-01-01T00:00:00.000Z",
          },
          "transactions": Array [
            Object {
              "amount": 1000,
              "category": null,
              "date": "2019-01-01T00:00:00.000Z",
              "description": "2 periods ago",
            },
          ],
        }
      }
    />
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="2"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-02-01T00:00:00.000Z",
            "start": "2019-01-15T00:00:00.000Z",
          },
          "transactions": Array [
            Object {
              "amount": 1000,
              "category": null,
              "date": "2019-01-15T00:00:00.000Z",
              "description": "1 period ago",
            },
          ],
        }
      }
    />
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="3"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-02-15T00:00:00.000Z",
            "start": "2019-02-01T00:00:00.000Z",
          },
          "transactions": Array [
            Object {
              "amount": 1000,
              "category": null,
              "date": "2019-02-06T00:00:00.000Z",
              "description": "Current period 1",
            },
            Object {
              "amount": 1000,
              "category": null,
              "date": "2019-02-08T00:00:00.000Z",
              "description": "Current period 2",
            },
          ],
        }
      }
    />
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="4"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-03-01T00:00:00.000Z",
            "start": "2019-02-15T00:00:00.000Z",
          },
          "transactions": Array [
            Object {
              "amount": 1000,
              "category": null,
              "date": "2019-02-26T00:00:00.000Z",
              "description": "Next period 2",
            },
          ],
        }
      }
    />
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="5"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-03-15T00:00:00.000Z",
            "start": "2019-03-01T00:00:00.000Z",
          },
          "transactions": Array [],
        }
      }
    />
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="6"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-04-01T00:00:00.000Z",
            "start": "2019-03-15T00:00:00.000Z",
          },
          "transactions": Array [],
        }
      }
    />
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="7"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-04-15T00:00:00.000Z",
            "start": "2019-04-01T00:00:00.000Z",
          },
          "transactions": Array [],
        }
      }
    />
    <TransactionsPeriodView
      className="transaction-view-period col-sm"
      key="8"
      period={
        Period {
          "openingBalance": 0,
          "range": t {
            "end": "2019-05-01T00:00:00.000Z",
            "start": "2019-04-15T00:00:00.000Z",
          },
          "transactions": Array [
            Object {
              "amount": 1000,
              "category": null,
              "date": "2019-04-30T00:00:00.000Z",
              "description": "In 5 periods",
            },
          ],
        }
      }
    />
  </div>
</div>
`;
